kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs

```diff
@@ -279,6 +282,385 @@ impl<'tcx> GotocCtx<'tcx> {
+                        );
+                    }
+                    Ordering::Greater => {
+                        // TODO: we also need to store the remaining bits by taking an offset _into
+                        // the same operand_ -- tests/kani/Slice/pathbuf.rs still passes for now,
+                        // but we are actually losing information here
+                        exprs.insert(
+                            comp.name(),
+                            operands_iter.next().unwrap().clone().reinterpret_cast(comp.typ()),
+                        );
+                    }
+                }
+            } else if niche_offset_bits < *target_offset_bits + comp_size_bits {
+                if comp.typ().is_struct_tag() {
+                    // recurse
+                    exprs.insert(
+                        comp.name(),
+                        self.codegen_niche_struct_from_operands_rec(
+                            operands_iter,
+                            operands_base_offset,
+                            niche_value_u128,
+                            niche_offset_bits,
+                            &comp.typ(),
+                            target_offset_bits,
+                        ),
+                    );
+                } else {
+                    exprs.insert(
+                        comp.name(),
+                        Expr::int_constant(
+                            niche_value_u128,
+                            Type::unsigned_int(comp.typ().sizeof_in_bits(&self.symbol_table)),
+                        )
+                        .transmute_to(comp.typ(), &self.symbol_table),
+                    );
+                }
+            }
+            *target_offset_bits += comp_size_bits;
+        }
+        Expr::struct_expr_with_nondet_fields(untagged_type.clone(), exprs, &self.symbol_table)
+    }
+
+    fn codegen_rvalue_enum(
+        &mut self,
+        variant_index: &VariantIdx,
+        operands: &[Operand<'tcx>],
+        ty: Ty<'tcx>,
+    ) -> Expr {
+        let overall_t = self.codegen_ty(ty);
+        let layout = self.layout_of(ty);
+        let adt = match &ty.kind() {
+            ty::Adt(adt, _) => adt,
+            _ => unreachable!(),
+        };
+        match &layout.variants {
+            Variants::Single { .. } => {
+                self.codegen_rvalue_enum_single(variant_index, operands, overall_t, adt)
+            }
+            Variants::Multiple { tag_encoding: TagEncoding::Direct, variants, .. } => {
+                self.codegen_rvalue_enum_direct(variant_index, operands, overall_t, adt, variants)
+            }
+            Variants::Multiple {
+                tag_encoding:
+                    TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start },
+                variants,
+                tag_field,
+                ..
+            } => {
+                // 1. Find the type of the member corresponding to the variant index.
```

Comment:
What if instead we do the following:
1. Create a temporary variable of the expected type.
2. Initialize the member of the corresponding variant.
3. Invoke the SetDiscriminant logic.
4. Return the temporary variable. 

Another possibility is to refactor the `codgen_rvalue` so it has access to the lvalue. So we wouldn't need the temporary variable. I believe that's closer to the rust backend logic.

Reply:
I now have an implementation of this proposal, and it indeed makes the implementation much easier (but might increase the burden on CBMC, which has to disentangle all the field updates).